dlt/destinations/path_utils.py

```diff
@@ -1,79 +1,276 @@
+from dlt.cli import echo as fmt
+from dlt.common.storages.load_package import ParsedLoadJobFileName
+from dlt.destinations.exceptions import CantExtractTablePrefix, InvalidFilesystemLayout
+from typing_extensions import Self
 
-# TODO: ensure layout only has supported placeholders
-SUPPORTED_PLACEHOLDERS = {"schema_name", "table_name", "load_id", "file_id", "ext", "curr_date"}
+from dlt.destinations.impl.filesystem.typing import TCurrentDateTime
 
-SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
 
+SUPPORTED_PLACEHOLDERS = {
+    "schema_name",
+    "table_name",
+    "load_id",
+    "file_id",
+    "ext",
+    "curr_date",
+    "year",
+    "month",
+    "day",
+    "hour",
+    "minute",
+    "dow",
+    "timestamp",
+}
 
-def check_layout(layout: str) -> List[str]:
-    placeholders = get_placeholders(layout)
-    invalid_placeholders = [p for p in placeholders if p not in SUPPORTED_PLACEHOLDERS]
-    if invalid_placeholders:
-        raise InvalidFilesystemLayout(invalid_placeholders)
-    return placeholders
+SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
 
 
 def get_placeholders(layout: str) -> List[str]:
     return re.findall(r"\{(.*?)\}", layout)
 
 
+class ExtraParams:
+    def __init__(
+        self,
+        current_datetime: TCurrentDateTime = None,
+        datetime_format: Optional[str] = None,
+        extra_placeholders: Optional[Dict[str, Any]] = None,
+        job_info: Optional[ParsedLoadJobFileName] = None,
+        schema_name: Optional[str] = None,
+        load_id: Optional[str] = None,
+    ) -> None:
+        self.current_datetime = current_datetime
+        self.datetime_format = datetime_format
+        self.extra_placeholders = extra_placeholders
+        self.job_info = job_info
+        self.load_id = load_id
+        self.schema_name = schema_name
+        self._params: Dict[str, Any] = {"schema_name": schema_name}
+
+        self.table_name = None
+        self.file_id = None
+        self.file_format = None
+        if job_info:
+            self.table_name = job_info.table_name
+            self.file_id = job_info.file_id
+            self.file_format = job_info.file_format
+            self._params.update(
+                {
+                    "table_name": self.table_name,
+                    "file_id": self.file_id,
+                    "ext": self.file_format,
+                }
+            )
+
+        if self.load_id:
+            self._params["load_id"] = self.load_id
+
+    @property
+    def params(self) -> Optional[Dict[str, Any]]:
+        """Process extra params for layout
+        If any value is a callable then we call it with the following arguments
+            * schema name,
+            * table name,
+            * load id,
+            * file id,
+            * current datetime
+        """
+        if not self.current_datetime:
+            self.current_datetime = pendulum.now()
```

Comment:
@rudolfix @sh-rp  athena also uses this helper utility should we also expose similar configuration to athena?

Reply:
yes - because athena must have all data for a particular table under a single folder. you should keep actual behavior of that utility and let filestystem (on replace) and athena to fail. pls make another push and I will give you more details